In [1]:
import pandas as pd 
import numpy as np

df = pd.read_csv("../data/crop_yield.csv")
df.head()

,State_Name,District_Name,Crop_Year,Season,Crop,Temperature,Humidity,Soil_Moisture,Area,Production
0,Andaman and Nicobar Islands,NICOBARS,2000,Kharif,Arecanut,36,35,45,1254.0,2000.0
1,Andaman and Nicobar Islands,NICOBARS,2000,Kharif,Other Kharif pulses,37,40,46,2.0,1.0
2,Andaman and Nicobar Islands,NICOBARS,2000,Kharif,Rice,36,41,50,102.0,321.0
3,Andaman and Nicobar Islands,NICOBARS,2000,Whole Year,Banana,37,42,55,176.0,641.0
4,Andaman and Nicobar Islands,NICOBARS,2000,Whole Year,Cashewnut,36,40,54,720.0,165.0


In [2]:
df.tail()

,State_Name,District_Name,Crop_Year,Season,Crop,Temperature,Humidity,Soil_Moisture,Area,Production
49994,Chhattisgarh,JANJGIR-CHAMPA,2006,Kharif,Small millets,36,35,45,27.0,7.0
49995,Chhattisgarh,JANJGIR-CHAMPA,2006,Kharif,Soyabean,37,40,46,27.0,32.0
49996,Chhattisgarh,JANJGIR-CHAMPA,2006,Kharif,Sunflower,36,41,50,450.0,213.0
49997,Chhattisgarh,JANJGIR-CHAMPA,2006,Kharif,Urad,37,42,55,1778.0,309.0
49998,Chhattisgarh,JANJGIR-CHAMPA,2006,Rabi,Gram,36,40,54,216.0,194.0


In [3]:
df.shape

(49999, 10)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49999 entries, 0 to 49998
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   State_Name     49999 non-null  object 
 1   District_Name  49999 non-null  object 
 2   Crop_Year      49999 non-null  int64  
 3   Season         49999 non-null  object 
 4   Crop           49999 non-null  object 
 5   Temperature    49999 non-null  int64  
 6   Humidity       49999 non-null  int64  
 7   Soil_Moisture  49999 non-null  int64  
 8   Area           49999 non-null  float64
 9   Production     49784 non-null  float64
dtypes: float64(2), int64(4), object(4)
memory usage: 3.8+ MB


In [5]:
df.isnull().sum()

State_Name         0
District_Name      0
Crop_Year          0
Season             0
Crop               0
Temperature        0
Humidity           0
Soil_Moisture      0
Area               0
Production       215
dtype: int64

In [6]:
set(df["Season"])

{'Autumn     ',
 'Kharif     ',
 'Rabi       ',
 'Summer     ',
 'Whole Year ',
 'Winter     '}

In [7]:
set(df["State_Name"])

{'Andaman and Nicobar Islands',
 'Andhra Pradesh',
 'Arunachal Pradesh',
 'Assam',
 'Bihar',
 'Chandigarh',
 'Chhattisgarh'}

In [8]:
#Chandigarh is not a state he is a capital  of panjab so replace Chandigarh -> panjab 
df["State_Name"] = df["State_Name"].replace({"Chandigarh":"Punjab"})

In [9]:
print("Chandigarh" in df.values)  # Should return False

False


In [10]:
print("Punjab" in df.values)  # Should return True

True


In [11]:
set(df["State_Name"])

{'Andaman and Nicobar Islands',
 'Andhra Pradesh',
 'Arunachal Pradesh',
 'Assam',
 'Bihar',
 'Chhattisgarh',
 'Punjab'}

In [12]:
df.groupby("State_Name")["District_Name"].nunique()

State_Name
Andaman and Nicobar Islands     3
Andhra Pradesh                 13
Arunachal Pradesh              18
Assam                          27
Bihar                          38
Chhattisgarh                   12
Punjab                          1
Name: District_Name, dtype: int64

In [13]:
df.loc[df["State_Name"]=="Bihar",["State_Name","District_Name"]].drop_duplicates()

,State_Name,District_Name
27005,Bihar,ARARIA
27571,Bihar,ARWAL
27868,Bihar,AURANGABAD
28315,Bihar,BANKA
28752,Bihar,BEGUSARAI
29342,Bihar,BHAGALPUR
29953,Bihar,BHOJPUR
30450,Bihar,BUXAR
30863,Bihar,DARBHANGA
31363,Bihar,GAYA


In [14]:
df.describe()

,Crop_Year,Temperature,Humidity,Soil_Moisture,Area,Production
count,49999.000000,49999.000000,49999.000000,49999.000000,49999.000000,4.978400e+04
mean,2005.879138,34.445489,44.776116,53.110502,7363.389469,4.132019e+05
std,5.106092,3.498524,6.662406,5.258764,27927.935789,1.232609e+07
min,1997.000000,25.000000,35.000000,45.000000,0.200000,0.000000e+00
25%,2002.000000,34.000000,40.000000,50.000000,76.000000,8.700000e+01
50%,2006.000000,36.000000,42.000000,54.000000,436.000000,6.384400e+02
75%,2010.000000,36.000000,50.000000,55.000000,2546.500000,5.615000e+03
max,2014.000000,37.000000,55.000000,62.000000,877029.000000,7.801620e+08


In [15]:
df.groupby("State_Name")["Soil_Moisture"].describe()

,count,mean,std,min,25%,50%,75%,max
State_Name,,,,,,,,
Andaman and Nicobar Islands,203.0,53.034483,5.267479,45.0,50.0,54.0,55.0,62.0
Andhra Pradesh,9628.0,53.110823,5.259514,45.0,50.0,54.0,55.0,62.0
Arunachal Pradesh,2546.0,53.112333,5.260442,45.0,50.0,54.0,55.0,62.0
Assam,14628.0,53.111362,5.258945,45.0,50.0,54.0,55.0,62.0
Bihar,18885.0,53.109717,5.258700,45.0,50.0,54.0,55.0,62.0
Chhattisgarh,4019.0,53.112963,5.258351,45.0,50.0,54.0,55.0,62.0
Punjab,90.0,53.111111,5.288199,45.0,50.0,54.0,55.0,62.0


hare we can directly see that in a soil_Moisture of a values min ,max,25%,50%,75% of the all state has as it is same data for all the state but each state has  a diff soil_moisture ,accoridng to the wather do that are giving  a incossistancy that's why we are droping a this columns 

In [16]:
df = df.drop("Soil_Moisture",axis=1)

hare yeild X area = production 
 we want total yeild  and we have 
 area and production 

 yeild = production /area

 

In [17]:
df["Yeild"] = (df["Production"]/df["Area"])

In [18]:
df.head()

,State_Name,District_Name,Crop_Year,Season,Crop,Temperature,Humidity,Area,Production,Yeild
0,Andaman and Nicobar Islands,NICOBARS,2000,Kharif,Arecanut,36,35,1254.0,2000.0,1.594896
1,Andaman and Nicobar Islands,NICOBARS,2000,Kharif,Other Kharif pulses,37,40,2.0,1.0,0.500000
2,Andaman and Nicobar Islands,NICOBARS,2000,Kharif,Rice,36,41,102.0,321.0,3.147059
3,Andaman and Nicobar Islands,NICOBARS,2000,Whole Year,Banana,37,42,176.0,641.0,3.642045
4,Andaman and Nicobar Islands,NICOBARS,2000,Whole Year,Cashewnut,36,40,720.0,165.0,0.229167


Area       → hectares (ha)


Production → tonnes (t)



Yield      → tonnes/hectare (t/ha)




In [19]:
df = df.drop(columns=["Crop_Year","Area","Production"])
df.head()

,State_Name,District_Name,Season,Crop,Temperature,Humidity,Yeild
0,Andaman and Nicobar Islands,NICOBARS,Kharif,Arecanut,36,35,1.594896
1,Andaman and Nicobar Islands,NICOBARS,Kharif,Other Kharif pulses,37,40,0.500000
2,Andaman and Nicobar Islands,NICOBARS,Kharif,Rice,36,41,3.147059
3,Andaman and Nicobar Islands,NICOBARS,Whole Year,Banana,37,42,3.642045
4,Andaman and Nicobar Islands,NICOBARS,Whole Year,Cashewnut,36,40,0.229167


we drop the ["Crop_Year","Area","Production"]) becouse of the this columns are calculated by 


        production = yeild * production 
and crop year are not the our feature 

In [20]:
df.isna().sum()

State_Name         0
District_Name      0
Season             0
Crop               0
Temperature        0
Humidity           0
Yeild            215
dtype: int64

hare yeild has the 215 empty val and yeild are a our label 
so we are droping a those val hare becouse if we replace them with a mean or madian  if craate a artificial output 
so thats why we are droing a that val 
''

In [21]:
df.dropna(subset=["Yeild"],inplace= True)

In [22]:
df.isnull().sum()

State_Name       0
District_Name    0
Season           0
Crop             0
Temperature      0
Humidity         0
Yeild            0
dtype: int64

In [23]:
df.shape

(49784, 7)

In [24]:
df.head()

,State_Name,District_Name,Season,Crop,Temperature,Humidity,Yeild
0,Andaman and Nicobar Islands,NICOBARS,Kharif,Arecanut,36,35,1.594896
1,Andaman and Nicobar Islands,NICOBARS,Kharif,Other Kharif pulses,37,40,0.500000
2,Andaman and Nicobar Islands,NICOBARS,Kharif,Rice,36,41,3.147059
3,Andaman and Nicobar Islands,NICOBARS,Whole Year,Banana,37,42,3.642045
4,Andaman and Nicobar Islands,NICOBARS,Whole Year,Cashewnut,36,40,0.229167


In [31]:
# now split the model into a treinig and testing dataset 
from sklearn.model_selection import train_test_split 

#we want to suffule dataset of the crop for testing so 
# we want a variours crop for thesting so we apply a staatify on a crop column to test   a various crop on a traind model
df = df[df["Crop"].map(df["Crop"].value_counts()) > 2]

train_set, test_set = train_test_split(df,
                                      stratify=df["Crop"],
                                      test_size=0.2,
                                      random_state=42
                                      )


In [33]:
train_set.shape

(39824, 7)

In [35]:
test_set.shape

(9957, 7)